lib/Mojolicious/Guides/Tutorial.pod

```diff
@@ -0,0 +1,938 @@
+  get '/headers' => sub ($c) {
+    my $url = $c->param('url') || 'https://mojolicious.org';
+    my $dom = $c->ua->get($url)->result->dom;
+    $c->render(json => $dom->find('h1, h2, h3')->map('text')->to_array);
+  };
+
+  # Non-blocking
+  get '/title' => sub ($c) {
+    $c->ua->get('mojolicious.org' => sub ($ua, $tx) {
+      $c->render(data => $tx->result->dom->at('title')->text);
+    });
+  };
+
+  # Concurrent non-blocking
+  get '/titles' => sub ($c) {
+    my $mojo = $c->ua->get_p('https://mojolicious.org');
+    my $cpan = $c->ua->get_p('https://metacpan.org');
+    Mojo::Promise->all($mojo, $cpan)->then(sub ($mojo, $cpan) {
+      $c->render(json => {
+        mojo => $mojo->[0]->result->dom->at('title')->text,
+        cpan => $cpan->[0]->result->dom->at('title')->text
+      });
+    })->wait;
+  };
+
+  app->start;
+
+For more information about the user agent see also L<Mojolicious::Guides::Cookbook/"USER AGENT">.
+
+=head2 WebSockets
+
+WebSocket applications have never been this simple before. Just receive messages by subscribing to events such as
+L<Mojo::Transaction::WebSocket/"json"> with L<Mojolicious::Controller/"on"> and return them with
+L<Mojolicious::Controller/"send">.
+
+  use Mojolicious::Lite -signatures;
+
+  websocket '/echo' => sub ($c) {
+    $c->on(json => sub ($c, $hash) {
+      $hash->{msg} = "echo: $hash->{msg}";
+      $c->send({json => $hash});
+    });
+  };
+
+  get '/' => 'index';
+
+  app->start;
+  __DATA__
+
+  @@ index.html.ep
+  <!DOCTYPE html>
+  <html>
+    <head>
+      <title>Echo</title>
+      <script>
+        const ws = new WebSocket('<%= url_for('echo')->to_abs %>');
+        ws.onmessage = function (event) {
+          document.body.innerHTML += JSON.parse(event.data).msg;
+        };
+        ws.onopen = function (event) {
+          ws.send(JSON.stringify({msg: 'I ♥ Mojolicious!'}));
+        };
+      </script>
+    </head>
+  </html>
+
+For more information about real-time web features see also L<Mojolicious::Guides::Cookbook/"REAL-TIME WEB">.
+
+=head2 Mode
+
+You can use the L<Mojo::Log> object from L<Mojolicious/"log"> to portably collect debug messages and automatically
+disable them later in a production setup by changing the L<Mojolicious> operating mode, which can also be retrieved
+from the attribute L<Mojolicious/"mode">.
+
+  use Mojolicious::Lite -signatures;
+
+  # Prepare mode specific message during startup
+  my $msg = app->mode eq 'development' ? 'Development!' : 'Something else!';
+
+  get '/' => sub ($c) {
+    $c->app->log->debug('Rendering mode specific message');
+    $c->render(text => $msg);
+  };
+
+  app->log->debug('Starting application');
+  app->start;
+
+The default operating mode will usually be C<development> and can be changed with command line options or the
+C<MOJO_MODE> and C<PLACK_ENV> environment variables. A mode other than C<development> will raise the log level from
+C<trace> to C<info>. All messages will be written to C<STDERR> by default.
```

Comment:
"All messages will be written to STDERR by default; however, this behavior is deactivated when a log file path is set."